### DISCUSSION TITLE
I2C problem with scanning and filestore

### DISCUSSION THREAD

**Discussion by james-drummond** [micropython/micropython#12284] [RP2040 / Pico] [answered]
PICO firmware:  MicroPython v1.20.0 on 2023-04-26;
I2C device:  one PCA9535  on the bus with 10k pull-up resistors
Using Thonny 4.1.1

The code is

```
from machine import Pin,I2C
i2c = I2C(0,scl=Pin(21),sda=Pin(20),freq=400000)
while True:
    i2c.scan()
print('Done')

```

If I put it into a file on my computer and then run it in Thonny - there is no scan result and no 'Done'
but
If I copy and paste the code directly into the Pico micropython interpreter then the scan results to [34] (the right number)

In both cases, I see traffic on the bus, and eyeballing it the traces look the same

I'm baffled.  Any help appreciated

**Comment by robert-hh** [comment, marked as answer]
That looks something caused by "Thonny-Magic". In it's attempt to be "user-friendly", Thonny does a lot of things under the hood, affecting the expected behavior. It should work better when you use e.g. mpremote.

Anyway, you should code `print(i2c.scan())` to see the scan result when the code is executed from a file.

**Comment by rkompass** [comment]
If you do the `i2c.scan()` in a while loop without wrapping a `print()` around it, it's return value, the array `[34]`, is not displayed.
If you do it in the REPL then you see the returned value, because the REPL always tries to print the result of the last command entered.

**Comment by james-drummond** [comment]
Thanks people I appreciate the help  - the main problem was in my brain.

Jim
